### PR TITLE
feat: update PodDecoration API design

### PR DESCRIPTION
**Pull request by Eikykun**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [ ] Y 

#### 2. What is the scope of this PR (e.g. component or file name):
PodDecoration Proposal

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by Eikykun** [comment]
Oth proposal,
1. Should PD select pods by collaset? such as
```yaml
  selector:               # Select the Pods to inject config.
    matchCollaSet:
    - collaSet-a
    - collaSet-b
    matchLabels:          # Select a group of Pods by labelSelector.
      ...
    matchExpressions:     # Or select specific Pods by matchExpressions.
      ...
```
2. RollingUpdate by CollaSet partition?
```yaml
  updateStrategy:
    rollingUpdate: 
      collaSet:
      - name: collaSet-a
         partition: 2 
```

cc @wu8685 @zoumo @shaofan-hs

File: docs/proposals/20231012-pod-decoration.md [hunk 1]
```diff
@@ -55,44 +54,59 @@ spec:
...
-          mountPath: /tmp/foo
-        targetSelectPolicy: AllAppContainers  # Indicates this configuration should be injected into all app containers.
-      volumes:
+      annotations:
+        extra-anno: anno-value
+    initContainers:                   
+    - name: init
+      image: init-docker:0.1.0
+    primaryContainers:               # Overwrite old container
+    - targetPolicy: ByName           # ByName | All | First | Last, default by name
```
Comment (wu8685) [inline docs/proposals/20231012-pod-decoration.md:67]: It would be better to use `All` as the default value. There usually is only one primary container.

**Comment by wu8685** [comment]
> Oth proposal,
> 
> 1. Should PD select pods by collaset? such as
> 
> ```yaml
>   selector:               # Select the Pods to inject config.
>     matchCollaSet:
>     - collaSet-a
>     - collaSet-b
>     matchLabels:          # Select a group of Pods by labelSelector.
>       ...
>     matchExpressions:     # Or select specific Pods by matchExpressions.
>       ...
> ```
> 
> 2. Support RollingUpdate by CollaSet partition?
> 
> ```yaml
>   updateStrategy:
>     rollingUpdate: 
>       collaSet:
>       - name: collaSet-a
>          partition: 2 
> ```
> 
> cc @wu8685 @zoumo @shaofan-hs

I think PD users need not to provide the CollaSet name. Telling PD which pods to inject could be enough.

About the item 2, I can't get your meaning. Why does PD need to know the CollaSet partition?

**Comment by Eikykun** [comment]
> > Oth proposal,
> > 
> > 1. Should PD select pods by collaset? such as
> > 
> > ```yaml
> >   selector:               # Select the Pods to inject config.
> >     matchCollaSet:
> >     - collaSet-a
> >     - collaSet-b
> >     matchLabels:          # Select a group of Pods by labelSelector.
> >       ...
> >     matchExpressions:     # Or select specific Pods by matchExpressions.
> >       ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. Support RollingUpdate by CollaSet partition?
> > 
> > ```yaml
> >   updateStrategy:
> >     rollingUpdate: 
> >       collaSet:
> >       - name: collaSet-a
> >          partition: 2 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > cc @wu8685 @zoumo @shaofan-hs
> 
> I think PD users need not to provide the CollaSet name. Telling PD which pods to inject could be enough.
> 
> About the item 2, I can't get your meaning. Why does PD need to know the CollaSet partition?

If PD selects Pods under multiple CollaSet, can I still use `updateStrategy.rollingUpdate.partition` to control the upgrade?
If so, how do I decide the pods under multiple CollaSets corresponding to the partition?

**Comment by wu8685** [comment]
> > I think PD users need not to provide the CollaSet name. Telling PD which pods to inject could be enough.
> > About the item 2, I can't get your meaning. Why does PD need to know the CollaSet partition?
> 
> If PD selects Pods under multiple CollaSet, can I still use `updateStrategy.rollingUpdate.partition` to control the upgrade? If so, how do I decide the pods under multiple CollaSets corresponding to the partition?

How about sharing the same data between CollaSet and PD controller? PD controller decides the Pod partition, and store this info into the shared memory. Then CollaSet controller is able to know the Pod partition info under each CollaSet from the shared memory.

Take the reference from this part of the [doc](https://github.com/KusionStack/operating/blob/3ad33e49ae3efe360ea7ffc8edfba95afff146fc/docs/proposals/20231012-pod-decoration.md#case-2-day-2-operation-canary-update-pod-additional-configuration)

cons: CollaSet controller can not run separately from PD controller

**Review by wu8685 — APPROVED** [review]
lgtm

**Pull request merged**